### ISSUE TITLE
inla.mode "twostage" log-posterior problem

### ISSUE THREAD

**Issue by finnlindgren** [hrue/r-inla#34]
The `$misc$configs` output appears to have incorrect values for `log.posterior` for each config for `inla.mode="twostage"`, with the output being `NaN` when `int.strategy="eb"`. There's also some interaction with the `num.threads` setting for `int.strategy="ccd"`, where the result is either `NaN` or values that do not agree with the other two `inla.mode` settings.

The problem was initially detected by an error message (seen below) from `inla.posterior.sample`, and appears for all models I've tested, including the simple intercept-only model in the example below, so it appears to be a universal issue.
The hyperparameter and effect estimates appear to be consistent across the`inla.mode` settings for the test example, so it's possible the error is only in the output and not in the values used internally for computing the posterior.

``` r
data <- data.frame(y = rnorm(4))
for (num.threads in list(NULL, "1:1")) {
  for (int.strategy in c("eb", "ccd")) {
    cat("\n--------------------------------------------\n")
    cat("num.threads = ", ifelse(is.null(num.threads), "NULL", num.threads), "\n",
        sep = "")
    cat("int.strategy = ", int.strategy, "\n", sep = "")
    fit <- list()
    for (mode in c("classic", "twostage", "experimental")) {
      fit[[mode]] <-
        INLA::inla(y ~ 1, data = data, family = "normal",
                   control.compute = list(config = TRUE),
                   control.inla = list(int.strategy = int.strategy),
                   inla.mode = mode,
                   num.threads = num.threads)
      
      cat("inla.mode = ", mode, "\n", sep = "")
      try({samp <- INLA::inla.posterior.sample(1, fit[[mode]])})
    }
    
    cat("config log.posterior values for int.strategy = ", int.strategy,
        ":\n", sep = "")
    print(
      as.data.frame(
        lapply(fit,
               function(x) vapply(x$misc$configs$config,
                                  function(y) {
                                    y$log.posterior
                                  },
                                  0.0)
        )
      )
    )
    cat("config log.posterior.orig values for int.strategy = ", int.strategy,
        ":\n", sep = "")
    print(
      as.data.frame(
        lapply(fit,
               function(x) vapply(x$misc$configs$config,
                                  function(y) {
                                    y$log.posterior.orig
                                  },
                                  0.0)
        )
      )
    )
  }
}
#> 
#> --------------------------------------------
#> num.threads = NULL
#> int.strategy = eb
#> inla.mode = classic
#> inla.mode = twostage
#> Error in sample.int(x, size, replace, prob) : NA in probability vector
#> inla.mode = experimental
#> config log.posterior values for int.strategy = eb:
#>   classic twostage experimental
#> 1       0      NaN            0
#> config log.posterior.orig values for int.strategy = eb:
#>   classic twostage experimental
#> 1       0      NaN            0
#> 
#> --------------------------------------------
#> num.threads = NULL
#> int.strategy = ccd
#> inla.mode = classic
#> inla.mode = twostage
#> Error in sample.int(length(x), size, replace, prob) : 
#>   NA in probability vector
#> inla.mode = experimental
#> config log.posterior values for int.strategy = ccd:
#>      classic twostage experimental
#> 1  0.0000000      NaN    0.0000000
#> 2 -0.6810429      NaN   -0.6810417
#> 3 -0.4589137      NaN   -0.4589142
#> config log.posterior.orig values for int.strategy = ccd:
#>      classic twostage experimental
#> 1  0.0000000      NaN    0.0000000
#> 2 -0.6810429      NaN   -0.6810417
#> 3 -0.4589137      NaN   -0.4589142
#> 
#> --------------------------------------------
#> num.threads = 1:1
#> int.strategy = eb
#> inla.mode = classic
#> inla.mode = twostage
#> Error in sample.int(x, size, replace, prob) : NA in probability vector
#> inla.mode = experimental
#> config log.posterior values for int.strategy = eb:
#>   classic twostage experimental
#> 1       0      NaN            0
#> config log.posterior.orig values for int.strategy = eb:
#>   classic twostage experimental
#> 1       0      NaN            0
#> 
#> --------------------------------------------
#> num.threads = 1:1
#> int.strategy = ccd
#> inla.mode = classic
#> inla.mode = twostage
#> inla.mode = experimental
#> config log.posterior values for int.strategy = ccd:
#>      classic  twostage experimental
#> 1  0.0000000 -2.274211    0.0000000
#> 2 -0.6808295 -2.733444   -0.6807928
#> 3 -0.4591909 -2.955006   -0.4592375
#> config log.posterior.orig values for int.strategy = ccd:
#>      classic  twostage experimental
#> 1  0.0000000  0.000000    0.0000000
#> 2 -0.6808295 -1.931733   -0.6807928
#> 3 -0.4591909 -2.153295   -0.4592375
```

<sup>Created on 2021-08-26 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by finnlindgren** [comment]
```
> sessionInfo()
R version 4.1.0 (2021-05-18)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 20.04.3 LTS

Matrix products: default
BLAS:   /home/flindgre/local/R-4.1.0/lib/R/lib/libRblas.so
LAPACK: /home/flindgre/local/R-4.1.0/lib/R/lib/libRlapack.so

locale:
 [1] LC_CTYPE=en_GB.UTF-8       LC_NUMERIC=C               LC_TIME=en_GB.UTF-8       
 [4] LC_COLLATE=en_GB.UTF-8     LC_MONETARY=en_GB.UTF-8    LC_MESSAGES=en_GB.UTF-8   
 [7] LC_PAPER=en_GB.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
[10] LC_TELEPHONE=C             LC_MEASUREMENT=en_GB.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] parallel  stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] INLA_21.07.10-1 sp_1.4-5        foreach_1.5.1   Matrix_1.3-4    reprex_2.0.1   
[6] usethis_2.0.1  

loaded via a namespace (and not attached):
 [1] lattice_0.20-44  codetools_0.2-18 withr_2.4.2      grid_4.1.0       lifecycle_1.0.0 
 [6] magrittr_2.0.1   rlang_0.4.11     fs_1.5.0         splines_4.1.0    iterators_1.0.13
[11] tools_4.1.0      glue_1.4.2       purrr_0.3.4      compiler_4.1.0 
```

**Comment by hrue** [comment]
fixed in commit c62026de491ebcaba8a2e3d03cd2cf24cb3ee916
